kolibri/content/utils/channel_import.py

```diff
@@ -376,33 +378,57 @@ def check_and_delete_existing_channel(self):
                 return False
-            self.check_cancelled()
-            root_id = ChannelMetadata.objects.get(pk=self.channel_id).root_id
-            ContentNodeClass = self.destination.get_class(ContentNode)
-            root_node = self.destination.session.query(ContentNodeClass).get(root_id)
-            # We set cascade behaviour on relationships on the default database
-            # So calling delete on the root node of a channel will delete everything else
-            self.destination.session.delete(root_node)
-            # Deletion/cascade is not enacted on the session until we call flush
-            # Note: as this function is generally called inside a explicitly managed
-            # transaction, this will not be committed to the database unless the entire
-            # update operation succeeds.
-            self.destination.session.flush()
+
         return True
 
+    def _can_use_optimized_pre_deletion(self):
+        # check whether we can skip fully deleting these three models, if we'll be using REPLACE on them anyway
+        for model in [ContentNode, File, AssessmentMetaData]:
+            mapping = self.schema_mapping.get(model, {})
+            row_mapper = self.generate_row_mapper(mapping.get('per_row'))
+            table_mapper = self.generate_table_mapper(mapping.get('per_table'))
+            if not self.can_use_sqlite_attach_method(model, row_mapper, table_mapper):
+                return False
+        return True
+
+    def delete_old_channel_data(self, old_tree_id):
+
+        # construct a template for deleting records for models that have a foreign key onto ContentNode
+        delete_related_template = "DELETE FROM {table} WHERE contentnode_id IN (SELECT id FROM content_contentnode WHERE tree_id = '{tree_id}')"
+
+        # construct a query for deleting old ContentNode records
+        delete_contentnode_query = "DELETE FROM content_contentnode WHERE tree_id = '{}'".format(old_tree_id)
+
+        if self._can_use_optimized_pre_deletion():
+            # if the external database is attached and there are no mappings for the relevant tables,
+            # we can skip deleting records that will be REPLACED during import, which helps efficiency
+            delete_related_template += " AND NOT id IN (SELECT id FROM sourcedb.{table})"
+            delete_contentnode_query += " AND NOT id IN (SELECT id FROM sourcedb.content_contentnode)"
+
+        # construct queries for deletion of File and AssessmentMetaData model records
+        delete_file_query = delete_related_template.format(table="content_file", tree_id=old_tree_id)
+        delete_assessment_query = delete_related_template.format(table="content_assessmentmetadata", tree_id=old_tree_id)
+
+        # run our deletion queries, starting with the related models (as they depend on the ContentNode records still existing)
+        for query in [delete_file_query, delete_assessment_query, delete_contentnode_query]:
```

Comment:
Will need to add queries to delete the M2M tables for ContentTag to ContentNode, and the ContentNode prerequisite, and related tables.